sklearn/grid_search.py

```diff
@@ -182,6 +182,103 @@ def __len__(self):
+                    cv, pre_dispatch, fit_params, iid, n_jobs, verbose):
+    """Actual fitting,  performing the search over parameters."""
+
+    estimator = clone(estimator)
+
+    if isinstance(scoring, (tuple, list)):
+        scorers = [check_scoring(estimator, scoring=s) for s in scoring]
+        ret_1d = False
+    else:
+        scorers = [check_scoring(estimator, scoring=scoring)]
+        ret_1d = True
+
+    n_samples = _num_samples(X)
+    X, y = check_arrays(X, y, allow_lists=True, sparse_format='csr')
+
+    if y is not None:
+        if len(y) != n_samples:
+            raise ValueError('Target variable (y) has a different number '
+                             'of samples (%i) than data (X: %i samples)'
+                             % (len(y), n_samples))
+        y = np.asarray(y)
+    cv = check_cv(cv, X, y, classifier=is_classifier(estimator))
+
+    if verbose > 0:
+        if isinstance(parameter_iterable, Sized):
+            n_candidates = len(parameter_iterable)
+            print("Fitting {0} folds for each of {1} candidates, totalling"
+                  " {2} fits".format(len(cv), n_candidates,
+                                     n_candidates * len(cv)))
+
+    out = Parallel(n_jobs=n_jobs, verbose=verbose, pre_dispatch=pre_dispatch)(
+            delayed(_fit_and_score)(
+                clone(estimator), X, y, scorers, train, test,
+                verbose, parameters, fit_params)
+            for parameters in parameter_iterable
+            for train, test in cv)
+
+    # `out` is a list of tuples (fold_scores, n_test, train_time, params).
+    n_fits = len(out)
+    n_folds = len(cv)
+    n_scorers = len(scorers)
+
+    grid_scores = []
+    for i in xrange(n_scorers):
+        grid_scores.append([])
+
+    for grid_start in range(0, n_fits, n_folds):
+
+        grid_stop = grid_start + n_folds
+        fold_scores, n_test, _, parameters = zip(*out[grid_start:grid_stop])
+        # `params` contains the same parameters n_fold times.
+        parameters = parameters[0]
+        # `fold_scores` is an n_folds x n_scorers 2-d array.
+        fold_scores = np.array(fold_scores)
+        weights = n_test if iid else None
+        mean_scores = np.average(fold_scores, axis=0, weights=weights)
+
+        for i in xrange(n_scorers):
+            # TODO: shall we also store the test_fold_sizes?
+            tup = _CVScoreTuple(parameters, mean_scores[i], fold_scores[:, i])
+            grid_scores[i].append(tup)
+
+    # Find the best parameters by comparing on the mean validation score:
+    # note that `sorted` is deterministic in the way it breaks ties
+    bests = [sorted(grid_scores[i], key=lambda x: x.mean_validation_score,
+                      reverse=True)[0] for i in xrange(n_scorers)]
+    best_params = [best.parameters for best in bests]
+    best_scores = [best.mean_validation_score for best in bests]
+
+    best_estimators = []
+    if refit:
+        for i in xrange(len(scorers)):
```

Comment:
I would use the first entry in the scoring list to find ``best_params`` and refit only for that. The other scores I'd just record.
That would not allow the change at runtime, as you suggested. However, is that a common use-case, or one that we should encourage? [I feel it encourages trying out multiple estimators on the test set]